### PR TITLE
Adds a react-todomvc app with the sam-pattern lib

### PR DESCRIPTION
**Pull request by jdubray**
The wiring of the pattern into the react component is in index.js 
The component itself (TodoApp.js) is wrapped into a component factoring that injects SAM's intents into the component and returns a hook to trigger the rendering once the SAM loop completes.

### CODE REVIEW

File: todomvc-app-react/src/index.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import '../node_modules/todomvc-app-css/index.css';
+import { TodoAppFactory, TodoApp }  from './sam/TodoApp';
+import * as serviceWorker from './serviceWorker';
+
+import initialState from './sam/initialState.js'
+import todo from './sam/todoComponent.js'
+// import { display } from './sam/display.js'
```
Comment (gunar) [inline todomvc-app-react/src/index.js]: deprecate comment?

File: todomvc-app-react/src/index.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import tp from 'sam-pattern'
+
+const { addInitialState, addComponent, setRender } = tp
+// wire it up  
+addInitialState(initialState)
+const { intents } = addComponent(todo)
+const todoApp = TodoAppFactory(intents, initialState)
+setRender(todoApp)
+
+// // start your application
```
Comment (gunar) [inline todomvc-app-react/src/index.js]: double `// //`

File: todomvc-app-react/src/index.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ReactDOM.render(<TodoApp />, document.getElementById('root'));
```
Comment (gunar) [inline todomvc-app-react/src/index.js]: This file should have the extension `.jsx` if you're using JSX

File: todomvc-app-react/src/sam/TodoApp.js [hunk 4]
```diff
@@ -0,0 +1,218 @@
+import React, { useState } from 'react';
+import '../../node_modules/todomvc-common/base.css'
+import '../../node_modules/todomvc-common/base.css'
+
+const clone = (state) => {
```
Comment (gunar) [inline todomvc-app-react/src/sam/TodoApp.js]: This one could use a comment. Very hard to read. No idea what it does.

File: todomvc-app-react/src/sam/TodoApp.js [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+          cln.__components.push(Object.assign(clone(c), { parent: cln }))
+        })
+      }
+    }
+    return cln
+  }
+
+export let TodoApp = () => {
+    return (<div/>)
+}
```
Comment (gunar) [inline todomvc-app-react/src/sam/TodoApp.js]: ?

**Comment by gunar** [comment]
The new library is looking great! When I write code I usually do a self-review before asking for others, because it's so easy to miss stuff—and I know you're a person that strives for high-quality.

File: todomvc-app-react/src/sam/TodoApp.js (hunk 4 above)
Comment (jdubray) [reply todomvc-app-react/src/sam/TodoApp.js]: There is a recursive relationship in the state (.parent), do you know if I still need to clone the state for React to work properly? I tried without cloning and it processes the first event/rendering correctly, then there is no more rendering. I thought the hooks were solving that problem.

File: todomvc-app-react/src/sam/TodoApp.js (hunk 5 above)
Comment (jdubray) [reply todomvc-app-react/src/sam/TodoApp.js]: Just to be safe if someone is trying to use the component before the factory has initialized it.

**Comment by jdubray** [comment]
@gunar thanks a lot! at the moment I am trying to see if the wiring I came up with is reasonable or if there is a better solution. Would have you done it differently?

**Comment by jdubray** [comment]
@gunar I moved cloning at the library level (via an option).

**Pull request merged**